### PR TITLE
Playcardfixes

### PR DESCRIPTION
**Pull request by MeszarosGabor**
- play_card() functionality added to Player. InvalidActionClass introduced for proper handling of invalid player actions. 
- I removed the player and other_card parameters from Card.action() as they can be retrieved at function call. On the other hand, I added other_player to emphasize that most cards require a target that needs to be provided.

### CODE REVIEW

File: love_letter_cards.py [hunk 1]
```diff
@@ -123,5 +125,11 @@ class Princess(Card):
...
         super().__init__("Princess", 8)
 
-    def action(self, player, other_card_in_hand):
+    def action(self, other_player):
         print("do Princess action")
+
+class InvalidActionError(Exception):
+    ''' Exception that is thrown when a player attempts an invalid action such as attacking another player protected by Handmaid.'''    
+    def __init__(self, message):
+        self.errorMessage = message
```
Comment (jima80525) [inline love_letter_cards.py:134]: Interesting.  I hadn't thought of that - mainly going down the "we won't pick invalid actions" path.  BUT this makes more sense when we get to the point where there are different ways to select the card to play and the opponent to play against.  Nice!

File: love_letter_sim2.py [hunk 2]
```diff
@@ -130,15 +131,18 @@ def more_than_one_player(self):
 
     def take_turn(self):
         player = next(self.turn_order)
+        player.protected = False # effect of a previously played Handmaid expires
```
Comment (jima80525) [inline love_letter_sim2.py:134]: Aha!  This fixes the problem I was trying to work around.  Didn't think to put that here.

File: love_letter_sim2.py [hunk 3]
```diff
@@ -80,27 +79,29 @@ def check_hand(self):
...
+        self.hand.pop(self.hand.index(played_card))
 
     def __str__(self):
-        if len(self.hand) == 1:
-            return "{0} holds {1}".format(self.id, self.hand[0])
-        else:
-            return "{0} holds {1} and {2}".format(self.id, self.hand[0],
-                                                  self.hand[1])
-
+        return  "{0} holds {1}".format(self.id, ' and '.join(map(str,self.hand)))
```
Comment (jima80525) [inline love_letter_sim2.py:104]: yep.  That's much better than the stupid thing I wrote. :)

File: love_letter_sim2.py [hunk 4]
```diff
@@ -80,27 +79,29 @@ def check_hand(self):
                         # JHA - we have to do something different here.
                         self.hand.remove(card)
 
-    def play_card(self,other_player = None):
+    def play_card(self,played_card = None,other_player = None):
```
Comment (jima80525) [inline love_letter_sim2.py:82]: I'm missing what played_card is for?  I'll point you to my repo where I've been working on this a bit....

File: love_letter_sim2.py [hunk 5]
```diff
@@ -130,15 +131,18 @@ def more_than_one_player(self):
...
+        player.protected = False # effect of a previously played Handmaid expires
         print("START TURN", player.id)
         if player.eliminated is False:
             # JHA - seems to me that the action a prince would have would be
             # to call 'discard()' on other player.  That player would be
             # responsible for playing card, and if princess: elminated =
             # True, else draw_card()
+            # GM: I agree, if two princes are played on the same player one after the other, the current check below is "too late".
             if len(player.hand) is 0:  # deal with Prince
                 player.draw_a_card(self.deck)
```
Comment (jima80525) [inline love_letter_sim2.py:143]: I still ended up taking this out.  If a prince is played against you, you discard what you have and, unless you're out, immediately draw a new card.  Waiting until now changes the deck order from the real game.

**Review by jima80525 — COMMENTED** [review]
Nice!

**Review by amosmj — APPROVED** [review]

**Pull request merged**